### DISCUSSION TITLE
Optional params?

### DISCUSSION THREAD

**Discussion by Maadtin** [TanStack/router#146] [Q&A] [answered]
In react router v6 they are no longer supported. 

for example `clients/:id?` would match both `/clients` and `/clients/1`

Is there a way to achieve something like this in react location?

**Comment by tannerlinsley** [comment, marked as answer]
Just use the same options for both the root route and param route.

Tanner Linsley

**Reply by Moshyfawn** [reply]
So, is this the idiomatic RL way to have an optional `param`? To duplicate the code for both **the root** and **the dynamic `param`** route? 
```jsx
{
  path: 'teams',
  children: [
    {
      path: '/',
      element: <Teams />
    },
    {
      path: ':id',
      element: <Teams />
    }
  ]
}
```

I'm looking for a solution to be able to get the optional parameter and, say, select a row with the selected data row `ID` from the URL. So, this way the same `<Teams />` component would render on both `teams` and `teams/:id` paths
```js
const { params: { id: teamId } } = useMatch()

useEffect(() => {
  if (teamId) {
    tableRef.current?.scrollToRow(teamId)
  }
}, [teamId])
```

[Here's an example CSB](https://stackblitz.com/edit/vitejs-vite-djnxcf?file=src/App.tsx)

**Reply by DCzajkowski** [reply]
I think this is the way, however this will cause `Teams` to fully unmount and remount during navigation between `/` and `/$id`. This can break focus etc.

I think the only way to do this without a full remount is to use [splat routes](https://tanstack.com/router/latest/docs/framework/react/guide/routing-concepts#splat--catch-all-routes).

**Comment by p7gg** [comment]
@tannerlinsley 
is there a built in for that in v1?

or I would have to declare 2 two routes, one with the param and one without it?
```ts
const route1  = new Route({
   path: 'user',
   ...
})

const route1  = new Route({
   path: 'user/$userID',
   ...
})
```

**Reply by tannerlinsley** [reply]
Correct :)

**Comment by DeluxeOwl** [comment]
@tannerlinsley Is there a plan to make this more ergonomic? Something similar to react-router https://reactrouter.com/en/main/route/route#optional-segments?

It's really annoying at the moment to support that exact pattern from the react-router link above, that is to have a lang param that sits at the root of your routes. 
```
/:lang?/some/really/long/path
```

You can declare two different routes, one for `/` and one for `/lang` - but that means you'd have to duplicate every other route that comes afterwards, which is very cumbersome (duplicate objects with different parent route if code base, otherwise duplicate files if the route is file-based). 


---

An alternative is to redirect on "/" to "/en" (if you're doing i18n), similarly to tanstack.com:
```
https://tanstack.com/router/ -> redirects to https://tanstack.com/router/latest - works ✅
https://tanstack.com/router/v1 - works ✅
```

That would work really nice for the docs as well
```
https://tanstack.com/router/latest/docs/latest/framework/react/overview - works ✅
https://tanstack.com/router/v1/docs/latest/framework/react/overview - works ✅

https://tanstack.com/router/docs/framework/react/overview - doesn't work ❌
```

**Reply by schiller-manuel** [reply]
eventually this is a feature that can be implemented, however it is not in our focus right now.
if you want to contribute via a PR, you are welcome!
also if you want to get this feature prioritized, reach out on discord to discuss sponsorship options.

**Comment by yslpn** [comment]
Our team is considering migrating from React Router to TanStack router, but this is the only blocker.

We have these paths

`/skill`
and
`/skill/$skillId`

We want to render the SkillPage component for two cases and type-safely use the parameters inside

```javascript
const skillId = useParams({
  from: `skill/$skillId`,
  select: (params) => params.skillId,
});
```

This is how we determine whether the skill is created or not.

**Reply by ilkergzlkkr** [reply]
you can use `strict: false` to get all possible params

```ts
const skillId = useParams({ strict: false }).skillId
// skillId: string | undefined
```

**Reply by yslpn** [reply]
I wanted to clarify a specific use case that highlights why this is problematic.

Imagine we have a route with **two parameters**, like this:

```
/account/$accountId/skill/$skillId?
```

In this case:
- `accountId` is **required**
- `skillId` is **optional**

Now, if I want to use `useParams()` in a type-safe way and only extract `accountId`, I might try to do this:

```ts
const accountId = useParams({
  from: "/account/$accountId/skill/$skillId",
  select: (params) => params.accountId,
});
```

But the problem is that because `skillId` is optional, I **must** include it in the `from` path to make this match. However, that now makes `accountId` **not type-safe anymore**, because the router can't know at compile time that it's **guaranteed** to be present.

The alternative is to use `strict: false`:

```ts
const { accountId } = useParams({ strict: false });
```

…but this gives me `accountId: string | undefined`, which isn't right — `accountId` is required in this route and should always exist.

The only way to get this type-safe would be to write complex conditional types or manually narrow it, which defeats the purpose of type safety and DX.

---

This limitation makes it really hard to work with routes that have a mix of required and optional params, while keeping the guarantees that TanStack Router usually provides.

**Reply by noudadrichem** [reply]
➕1 on having optional parameters in routes. We're migrating from react-router to TanStack Router as well because of its preloading functionality. But we have to create multiple routes for the same components to get optional route params to work. 

Feels a bit weird.

**Reply by tannerlinsley** [reply]
We’re almost there.

**Reply by medihack** [reply]
Is there some issue or PR to follow? (We would like to use this feature for i18n routing.)

**Reply by Amixx** [reply]
Seconding medihack - is there a PR to follow? Looking forward to this feature!

**Reply by ilarramendi** [reply]
Also interested in following the process of this feature!

**Reply by kimwes** [reply]
I was also interested on this feature and looked a bit. At least to me it seems that there has been some work on the devinxi-parse-path-params branch that may be related to the optional path params. Probably needs the whole devinxi to be done first.

**Reply by gijsmin** [reply]
For me the same, wanting to migrate to tanstack router but really need this optional path params

**Reply by rek** [reply]
Seems like it is tracked by: https://github.com/orgs/TanStack/projects/4?pane=issue&itemId=86817540

**Reply by masterbater** [reply]
it is already supported check out the docs

**Reply by masterbater** [reply]
https://tanstack.com/router/v1/docs/framework/react/guide/path-params#how-optional-parameters-work

**Reply by hyusetiawan** [reply]
@masterbater do you know how to do this with file based routing? for ex: `user.$id.tsx` i can't figure out how to make the $id optional

**Reply by masterbater** [reply]
> @masterbater do you know how to do this with file based routing? for ex: `user.$id.tsx` i can't figure out how to make the $id optional

reading from the docs just add hyphen and curly braces to $id

user.{-$id}.tsx if you do flat routes

// Single optional parameter
// src/routes/posts/{-$category}.tsx

**Reply by mkarajohn** [reply]
I cannot figure this out on non-flat file based routing.

I tried writing the route like this: `src/routes/_authenticated/applications/$applicationId/{-$checkGroups}/{-$check}/index.tsx` (with `$applicationId` and `{-$checkGroups}` directories empty) and I get `Invariant failed: Could not find an active match from "/_authenticated/applications/$applicationId/"` whenever I go anywhere beyond ``applications/`

I tried copy pasting the same index file in all the in-between segments (which is not very ergonomic, so not sure this is the correct way), like this:

```
frontend/src/routes/_authenticated/applications/$applicationId/
├── {-$checkGroups}
│   ├── {-$check}
│   │   └── index.tsx
│   └── index.tsx
└── index.tsx
``` 

and i get `
Invariant failed: Could not find an active match from "/_authenticated/applications/$applicationId/"` when i go anywhere beyond `applications/$applicationId/`

So, how does this work, it's not clear, can someone explain?

**Reply by schiller-manuel** [reply]
can you please create a new GitHub issue including a complete minimal reproducer ?

**Reply by mkarajohn** [reply]
Before I do that, would you be able to point me to usage documentation of this feature on non-flat file based routing? I think this is a documentation issue; in which case we only need to update that.

**Reply by mkarajohn** [reply]
UPDATE: i found the issue: I had a couple forgotten `useParams` that I never updated the `from` argument :sweat_smile: 

*however* I still think the docs on how to use this with non-flat file routing are lacking, I'll see if I have time to open a PR to add something in the docs

UPDATE2: in hindsight, and upon further introspection, this is probably a skill issue; I may have to take a better look to the file based routing docs. Thank you for coming to my TED talk

**Comment by masterbater** [comment]
For someone needed a  updated answer it is already supported

https://tanstack.com/router/v1/docs/framework/react/guide/path-params#defining-optional-parameters


Optional parameters create flexible URL patterns:

/posts/{-$category} matches both /posts and /posts/tech
/posts/{-$category}/{-$slug} matches /posts, /posts/tech, and /posts/tech/hello-world
/users/$id/{-$tab} matches /users/123 and /users/123/settings
When an optional parameter is not present in the URL, its value will be undefined in your route handlers and components.

**Reply by rek** [reply]
[its only 1 week old](https://github.com/TanStack/router/pull/4597), but thanks for the update here!!

**Reply by gijsmin** [reply]
Indeed its just released
